algebra/src/fft/domain/basic_radix_2_domain.rs

```diff
@@ -0,0 +1,286 @@
+    pub size_as_field_element: F,
+    /// Inverse of the size in the field.
+    pub size_inv:              F,
+    /// A generator of the subgroup.
+    pub group_gen:             F,
+    /// Inverse of the generator of the subgroup.
+    pub group_gen_inv:         F,
+    /// Multiplicative generator of the finite field.
+    pub generator_inv:         F,
+}
+
+impl<F: PrimeField> fmt::Debug for BasicRadix2Domain<F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Multiplicative subgroup of size {}", self.size)
+    }
+}
+
+impl<F: PrimeField> BasicRadix2Domain<F> {
+
+    /// Given an index which assumes the first elements of this domain are the elements of
+    /// another (sub)domain with size size_s,
+    /// this returns the actual index into this domain.
+    pub fn reindex_by_subdomain(&self, other: Self, index: usize) -> usize {
+        assert!(self.size() >= other.size());
+        // Let this subgroup be G, and the subgroup we're re-indexing by be S.
+        // Since its a subgroup, the 0th element of S is at index 0 in G, the first element of S is at
+        // index |G|/|S|, the second at 2*|G|/|S|, etc.
+        // Thus for an index i that corresponds to S, the index in G is i*|G|/|S|
+        let period = self.size() / other.size();
+        if index < other.size() {
+            index * period
+        } else {
+            // Let i now be the index of this element in G \ S
+            // Let x be the number of elements in G \ S, for every element in S. Then x = (|G|/|S| - 1).
+            // At index i in G \ S, the number of elements in S that appear before the index in G to which
+            // i corresponds to, is floor(i / x) + 1.
+            // The +1 is because index 0 of G is S_0, so the position is offset by at least one.
+            // The floor(i / x) term is because after x elements in G \ S, there is one more element from S
+            // that will have appeared in G.
+            let i = index - other.size();
+            let x = period - 1;
+            i + (i / x) + 1
+        }
+    }
+
+    fn best_fft(a: &mut [F], worker: &Worker, omega: F, log_n: u32) {
+        let log_cpus = worker.log_num_cpus();
+
+        if log_n <= log_cpus {
+            Self::serial_fft(a, omega, log_n);
+        } else {
+            Self::parallel_fft(a, worker, omega, log_n, log_cpus);
+        }
+    }
+
+    /// Computes the radix-2 FFT of a[0..n] over an FFT domain {z: z^n - 1 = 0} of size n=2^log_n, 
+    /// given a generator  omega for this domain.
+    ///
+    /// The algorithm reindexes the FFT domain C_n={0,1}^log(n) by reversing the bit order. 
+    /// This makes the enumeration of the FFT domain C_n into the cosets of C_m with m|n slightly
+    /// more intuitive: For fixed k' from 0..n/m, 
+    ///              [k' || j] = k' * m + j, 
+    /// where j=0..m, goes through the coset of C_m at k', and varying k' enumerates the partition
+    /// of C_n.
```

Comment:
Actually it seems that "inverse coordinates" are essential to work with one and the same data structure (i.e. the state vector). That needs to be better understood if we want to generalize to the general mixed radix case with arbitrary number of prime factors.